File: Array/59-Spiral-Matrix-II/59-Spiral-Matrix-II.java
class Solution {
    public int[][] generateMatrix(int n) {
        int[][] ans = new int[n][n];

        int left = 0, right = n;
        int top = 0, bottom = n;
        int num = 1;

        while(top < bottom && left < right) {
            for(int i = left;i<right;i++) {
                ans[top][i] = num++;
            }
            top++;

            for(int i=top;i<bottom;i++){
                ans[i][right - 1] = num++;
            }
            right--;

            for(int i=right-1;i>=left;i--) {
                ans[bottom-1][i] = num++;
            }
            bottom--;

            for(int i=bottom-1;i>=top;i--) {
                ans[i][left] = num++;
            }
            left++;
        }
        return ans;
    }
}

File: Array/59-Spiral-Matrix-II/59-Spiral-Matrix-II.js
/**
 * @param {number} n
 * @return {number[][]}
 */
var generateMatrix = function(n) {
        let ans = [...new Array(n)].map(e => [...new Array(n)].fill(0));

        let left = 0, right = n;
        let top = 0, bottom = n;
        let num = 1;

        while(top < bottom && left < right) {
            for(let i = left;i<right;i++) {
                ans[top][i] = num++;
            }
            top++;

            for(let i=top;i<bottom;i++){
                ans[i][right - 1] = num++;
            }
            right--;

            for(let i=right-1;i>=left;i--) {
                ans[bottom-1][i] = num++;
            }
            bottom--;

            for(let i=bottom-1;i>=top;i--) {
                ans[i][left] = num++;
            }
            left++;
        }
        return ans;
};
